### PR TITLE
[Berkeley] Evaluate optional gate selectors

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds handling for evaluations of the optional gate selectors in pickles, as part of the work to remove the error-prone, expensive, and unwieldy linearization part of kimchi proofs. The proof-systems counterpart can be found in https://github.com/o1-labs/proof-systems/pull/1109.

As part of these changes, the `Plonk_types.Opt.Early_stop_sequence.fold` for absorbing commitments was removed, and replaced with a lightweight-but-less-flexible version of the `Opt_sponge`. This code is currently inline, pending a refactor in an upcoming PR.

This sponge change also fixed another underlying bug: the 'runtime selector' evaluation was never absorbed correctly unless all of the possible columns of the lookup argument were required for a custom gate. This was an oversight in our testing, and will be exercised by the final end-to-end-test.

This PR builds upon https://github.com/MinaProtocol/mina/pull/13531.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/pickles_types/plonk_types.ml [hunk 1]
```diff
@@ -414,7 +414,7 @@ module Evals = struct
     let opt_typ impl ({ lookup; runtime_tables; _ } : Opt.Flag.t Features.t)
         ~dummy:z elt =
       Opt.typ impl lookup
-        ~dummy:(dummy z ~runtime:(not (Opt.Flag.equal runtime_tables No)))
+        ~dummy:(dummy z ~runtime:(Opt.Flag.equal runtime_tables Yes))
```
Comment (rbonichon) [inline src/lib/pickles_types/plonk_types.ml:417]: This is actually a bugfix since it changes the behavior for `Maybe`

**Review by rbonichon — APPROVED** [review]
LGTM

File: src/lib/pickles_types/plonk_types.ml (hunk 1 above)
Comment (querolita) [reply src/lib/pickles_types/plonk_types.ml:417]: Interesting point

**Review by querolita — APPROVED** [review]

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**